### PR TITLE
DONE: Add Sorting to InventoryModel

### PR DESCRIPTION
**Pull request by Choo-Xing-Yu**
Items are now sorted by their % maxQuantity.
If items do not have % maxQuantity they are flushed to the end of the list and sorted by their Quantity.
Items with the same % maxQuantity are sorted lexicographically by their name.

### CODE REVIEW

File: src/main/java/seedu/address/model/item/comparator/ItemComparator.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        double percentageQuantity2 = maxQuantity2.map(maxQ -> quantity2.divideBy(maxQ))
+                .orElse(Double.MAX_VALUE);
+
+        if (percentageQuantity1 < percentageQuantity2) {
+            return -1;
+        } else if (percentageQuantity1 > percentageQuantity2) {
+            return 1;
+        } else {
+            return 0;
+        }
```
Comment (saad-haider-pp) [inline src/main/java/seedu/address/model/item/comparator/ItemComparator.java]: In the else condition, you should return quantity1 - quantity2

i.e. when percentage is same, or when there is no percentage, then the item with lesser quantity is higher up

**Comment by Wincenttjoi** [comment]
@xnoobftw If you are okay with implementing, I think it would be good to return a message on the commandbox to make it consistent with other commands

File: src/main/java/seedu/address/model/item/comparator/ItemComparator.java (hunk 1 above)
Comment (Choo-Xing-Yu) [reply src/main/java/seedu/address/model/item/comparator/ItemComparator.java]: Done! Take a look

**Comment by Choo-Xing-Yu** [comment]
> @xnoobftw If you are okay with implementing, I think it would be good to return a message on the commandbox to make it consistent with other commands

May i know what you mean by this?

**Comment by Wincenttjoi** [comment]
> > @xnoobftw If you are okay with implementing, I think it would be good to return a message on the commandbox to make it consistent with other commands
> 
> May i know what you mean by this?

So an example would be after the command `sort ...` a line should return on the command box
image
A possible reply would be like: You inventory item/delivery is sorted according to .............

**Comment by Choo-Xing-Yu** [comment]
> > > @xnoobftw If you are okay with implementing, I think it would be good to return a message on the commandbox to make it consistent with other commands
> > 
> > 
> > May i know what you mean by this?
> 
> So an example would be after the command `sort ...` a line should return on the command box
> image
> A possible reply would be like: You inventory item/delivery is sorted according to .............

If i'm not wrong our agreement was that sorting will be done implicitly and does not require a sort command

File: src/main/java/seedu/address/logic/Logic.java [hunk 2]
```diff
@@ -37,6 +38,9 @@
     /** Returns an unmodifiable view of the filtered list of items */
     ObservableList<Item> getFilteredItemList();
 
+    /** Returns an unmodifiable view of the sorted list of items */
+    SortedList<Item> getSortedItemList();
+
```
Comment (saad-haider-pp) [inline src/main/java/seedu/address/logic/Logic.java:39]: Do we want a separate (from `getFilteredItemList()`) method here? Why not let sorting be done within `getFilteredItemList()`

File: src/main/java/seedu/address/model/inventorymodel/InventoryModelManager.java [hunk 3]
```diff
@@ -143,6 +139,11 @@ public void updateFilteredItemList(Predicate<Item> predicate) {
         filteredItems.setPredicate(predicate);
     }
 
+    @Override
+    public SortedList<Item> getSortedItemList() {
+        return new SortedList<Item>(filteredItems, ITEM_COMPARATOR);
+    }
```
Comment (saad-haider-pp) [inline src/main/java/seedu/address/model/inventorymodel/InventoryModelManager.java]: why not return this in `getFilteredItemList()`

File: src/main/java/seedu/address/model/item/Name.java [hunk 4]
```diff
@@ -56,4 +56,8 @@ public int hashCode() {
         return fullName.hashCode();
     }
 
+    @Override
+    public int compareTo(Name o) {
+        return fullName.compareTo(o.fullName);
+    }
```
Comment (saad-haider-pp) [inline src/main/java/seedu/address/model/item/Name.java:62]: Not an issue, just that this is the first time I have seen o as a variable name 😆

Comment (Choo-Xing-Yu) [reply src/main/java/seedu/address/model/item/Name.java:62]: o it's auto generated by intellj HAHA but i'll change it to match!

File: src/main/java/seedu/address/model/inventorymodel/InventoryModelManager.java (hunk 3 above)
Comment (Choo-Xing-Yu) [reply src/main/java/seedu/address/model/inventorymodel/InventoryModelManager.java]: The method name would be misleading in my opinion

File: src/main/java/seedu/address/logic/Logic.java (hunk 2 above)
Comment (xnoobftw) [reply src/main/java/seedu/address/logic/Logic.java:40]: Good idea! Done that please check!

Comment (xnoobftw) [reply src/main/java/seedu/address/logic/Logic.java:39]: Edit bugs out if we were to do that!

Comment (Choo-Xing-Yu) [reply src/main/java/seedu/address/logic/Logic.java:39]: Attempting to do it now

Comment (Choo-Xing-Yu) [reply src/main/java/seedu/address/logic/Logic.java:39]: Done! @halcon-blanco please check

File: src/main/java/seedu/address/model/inventorymodel/InventoryModelManager.java (hunk 3 above)
Comment (Choo-Xing-Yu) [reply src/main/java/seedu/address/model/inventorymodel/InventoryModelManager.java]: @halcon-blanco did that! it turns out this helped iron out bugs so much thanks!

File: src/main/java/seedu/address/logic/commands/itemcommand/ItemRemoveCommand.java [hunk 5]
```diff
@@ -59,9 +59,9 @@ public CommandResult execute(Model model) throws CommandException {
         requireInventoryModel(model);
         InventoryModel inventoryModel = (InventoryModel) model;
 
-        List<Item> lastShownList = inventoryModel.getFilteredItemList();
+        List<Item> lastShownList = inventoryModel.getFilteredAndSortedItemList();
 
-        if (index.getZeroBased() >= lastShownList.size()) {
+        if (index.getZeroBased() >= lastShownList.size() && index.getZeroBased() < 0) {
             throw new CommandException(Messages.MESSAGE_INVALID_ITEM_DISPLAYED_INDEX);
         }
```
Comment (saad-haider-pp) [inline src/main/java/seedu/address/logic/commands/itemcommand/ItemRemoveCommand.java:66]: should it be `&&` or `||`

File: src/main/java/seedu/address/model/inventorymodel/InventoryModelManager.java [hunk 6]
```diff
@@ -126,19 +122,19 @@ public void setItem(Item target, Item editedItem) {
...
+    public ObservableList<Item> getFilteredAndSortedItemList() {
+        return new SortedList<>(filteredItems, ITEM_COMPARATOR);
     }
 
     @Override
-    public void updateFilteredItemList(Predicate<Item> predicate) {
+    public void updateFilteredAndSortedItemList(Predicate<Item> predicate) {
         requireNonNull(predicate);
         filteredItems.setPredicate(predicate);
     }
```
Comment (saad-haider-pp) [inline src/main/java/seedu/address/model/inventorymodel/InventoryModelManager.java:140]: I don't think the "Sorted" part is accurate here. The old name `updateFilteredItemList` sounds more appropriate

File: src/main/java/seedu/address/model/item/Quantity.java [hunk 7]
```diff
@@ -83,4 +92,8 @@ public int hashCode() {
         return value.hashCode();
     }
 
+    @Override
+    public int compareTo(Quantity o) {
+        return Double.compare(Double.parseDouble(value), Double.parseDouble(o.value));
+    }
 }
```
Comment (saad-haider-pp) [inline src/main/java/seedu/address/model/item/Quantity.java:99]: Sorry this one very nitpicky, but if you're changing the "o's", can you change this one too

File: src/main/java/seedu/address/model/item/comparator/ItemNameComparator.java [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+
+import java.util.Comparator;
+
+import seedu.address.model.item.Item;
+
+public class ItemNameComparator implements Comparator<Item> {
+    @Override
+    public int compare(Item o1, Item o2) {
+        return o1.getName().compareTo(o2.getName());
+    }
```
Comment (saad-haider-pp) [inline src/main/java/seedu/address/model/item/comparator/ItemNameComparator.java:11]: and this one

File: src/main/java/seedu/address/model/inventorymodel/InventoryModelManager.java (hunk 6 above)
Comment (Choo-Xing-Yu) [reply src/main/java/seedu/address/model/inventorymodel/InventoryModelManager.java:140]: I changed it to updateItemListFilter!

File: src/main/java/seedu/address/logic/commands/itemcommand/ItemRemoveCommand.java (hunk 5 above)
Comment (Choo-Xing-Yu) [reply src/main/java/seedu/address/logic/commands/itemcommand/ItemRemoveCommand.java:66]: Opps my bad! Thanks for pointing it out

File: src/main/java/seedu/address/model/item/comparator/ItemNameComparator.java (hunk 8 above)
Comment (Choo-Xing-Yu) [reply src/main/java/seedu/address/model/item/comparator/ItemNameComparator.java:11]: Changed!

File: src/main/java/seedu/address/model/item/Quantity.java (hunk 7 above)
Comment (Choo-Xing-Yu) [reply src/main/java/seedu/address/model/item/Quantity.java:99]: Changed!

**Review by saad-haider-pp — APPROVED** [review]
LGTM! Great job!

**Pull request merged**